package robogenMessage;

message EvolvableParameter {
   required string paramName = 1;
   required float paramValue = 2;
}

message BodyPart {
  required string id = 1;
  required string type = 2;
  required bool root = 3;
  repeated EvolvableParameter evolvableParam = 4;
  required int32 orientation = 5;
}

message BodyConnection {
  required string src = 1; 
  required string dest = 2;
  required int32 srcSlot = 3;
  required int32 destSlot = 4;
}

message Body {
  repeated BodyPart part = 1;
  repeated BodyConnection connection = 2;
}

message NeuralConnection {
  required string src = 1; 
  required string dest = 2;
  required float weight = 3;
}

message Neuron {
  required string id = 1;
  required string layer = 2;
  optional float biasWeight = 3;
  optional string bodyPartId = 4; 
  optional int32 ioId = 5;
}

message Brain {
  repeated Neuron neuron = 1;
  repeated NeuralConnection connection = 2;
}

message Robot {
  required int32 id = 1;
  required Body body = 2; 
  required Brain brain = 3;
  required string configuration = 4;
}

message Obstacle {
  required float x = 1;
  required float y = 2;
  required float z = 3;
  required float xSize = 4;
  required float ySize = 5;
  required float zSize = 6;
  required float density = 7;
}

message StartPosition {
  required float x = 1;
  required float y = 2;
  required float azimuth = 3;
}

message RobogenConfig {
  required float terrainWidth = 1;
  required float terrainLength = 2;
  repeated Obstacle obstacles = 3;
  required string scenario = 4;
  required float lightSourceHeight = 5;
  required float timeStep = 6;
  required int32 nTimeSteps = 7;
  repeated StartPosition startPositions = 8;
}

message Task {
  required Robot robot = 1;
  required RobogenConfig conf = 2;
}

message EvaluationResult {
    required int32 id = 1; 
    required float fitness = 2;
    repeated float objectives = 3;
}

